### PR TITLE
1859569: Abort on invalid username/token argument in syspurpose subcommands

### PR DESCRIPTION
**Pull request by m-horky**
* Card ID: ENT-2697
* BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1859569

- If the system is registered, abort if --username, --password, --token or --org are specified.
- In man page, the --org argument has been moved up so it is directly below other arguments usable on unregistered systems.
- For those arguments the description has been altered to clarify their meaning.

### CODE REVIEW

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (jirihnidek) [inline src/subscription_manager/cli_command/abstract_syspurpose.py:165]: Linguistic question: Wouldn't be better to use "Error: --username, --password and --token can be used only on unregistered system." ? @tlhmerry9 @cnsnyder Thoughts?

File: test/cli_command_test/test_role.py
Comment (jirihnidek) [inline test/cli_command_test/test_role.py:106]: I would add more unit tests for other cases (password, org and token).

File: test/cli_command_test/test_role.py
Comment (jirihnidek) [inline test/cli_command_test/test_role.py:114]: I would expect to see here `self.cc.options.list = True`.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall I like the PR. I have only few small requests.

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (ptoscano) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:165]: @jirihnidek as non-English speaker: from what I know, there is a small difference in the meaning of the two versions (the current and yours):
- "it can only be used on ..." == "the ability [can] applies only on the usability of that on ..."
- "it can be used only on ..." == "you can use ... only on"

Again, AFAIK. In the end, I think the general meaning of both should be the same in this context.

File: man/subscription-manager.8
Comment (ptoscano) [inline man/subscription-manager.8:432]: if the subject of "applies" is "addons", then it ought to be "apply"

File: test/cli_command_test/test_role.py
Comment (m-horky) [reply test/cli_command_test/test_role.py:114]: The new code in `_validate_options()` does not check if the `--list` argument is set or not. How is this different from setting it to `None` or not having it specified at all?

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (jirihnidek) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:165]: OK :smiley:

File: test/cli_command_test/test_role.py
Comment (jirihnidek) [reply test/cli_command_test/test_role.py:114]: That's true, but we try to mimic role command with `--list` option as much as possible. It is possible that somebody will add check to `_validate_option()` if the `--list` option is set or not in the future. In that case your unit test could give false positive result. When you are coding something it is also important to look into the future and consider future risks not only current state.

**Review by ptoscano — APPROVED** [review]
LGTM, thanks!

@jirihnidek, do you want to give this PR another look?

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (cnsnyder) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:166]: I think it is fine as is, with one small change "Error: --username, --password and --token can be used only on an unregistered system."

Comment (m-horky) [reply src/subscription_manager/cli_command/abstract_syspurpose.py]: Fixed.

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (cnsnyder) [inline src/subscription_manager/cli_command/abstract_syspurpose.py:158]: Can we implement this for the service-level command as well? (In a separate commit)

**Review by cnsnyder — APPROVED** [review]
Looks good to me and sounds like this aligns with stakeholders interests.

I think we can address the removal of the username and password usage in 'service-level' in another commit or PR, I'll leave this open for you to have the opportunity to include it in this PR.
I think it is worth merging as it is as well.

Thanks for the great work!

**Comment by jirihnidek** [comment]
> Looks good to me and sounds like this aligns with stakeholders interests.
> 
> I think we can address the removal of the username and password usage in 'service-level' in another commit or PR, I'll leave this open for you to have the opportunity to include it in this PR.
> I think it is worth merging as it is as well.
> 

I thought that we agreed to implement it in different way (to enable using --username and --password, when system is registered to be consistent with other commands). Did anything change?

**Comment by m-horky** [comment]
> I thought that we agreed to implement it in different way (to enable using --username and --password, when system is registered to be consistent with other commands). Did anything change?

We did, but I don't think this PR is contradicting that idea.

These options are currently silently ignored. In my opinion it makes sense to give feedback to the user that their arguments are invalid. When the --username/... arguments are supported on registered systems, this check will be removed. I see it as temporary  clarification, not permanent solution.

File: src/subscription_manager/cli_command/abstract_syspurpose.py
Comment (m-horky) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:158]: Why in a separate commit and not in a single one? service-level is syspurpose subcommand, and has been left out kind of by accident, because it implements its own validation.

Comment (cnsnyder) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:158]: It could be in the same commit if preferred.

**Comment by m-horky** [comment]
The check has been duplicated to `service-level` verification, too. The same goes for tests for it.

**Review by cnsnyder — APPROVED** [review]

**Comment by cnsnyder** [comment]
Looks good to me. Also got QE signoff on the changes, although test changes will be required (not in the unit tests but in QE's tests). Will merge this after the PR for subscription-manager-1.28 is opened.

**Comment by cnsnyder** [comment]
Merging following the desire of PM

**Comment by m-horky** [comment]
Following the conversation in 1.28 PR #2737 the message has been changed:
```diff
- "can be used only on unregistered system."
+ "can be used only on unregistered systems."
```

**Review by ptoscano — APPROVED** [review]
LGTM now; looked good for Chris; Jiri's notes were fixed.

Hence, accepting it.

**Pull request merged**
